### PR TITLE
#167: added example

### PR DESCRIPTION
**Pull request by goto40**

### CODE REVIEW

File: tests/functional/regressions/test_issue167.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+def person_definer_scope(knows, attr, attr_ref):
+    m = get_model(knows)  # get the model of the currently processed element
+    name = attr_ref.obj_name  # the name of currently looked up element
+    found_persons = list(filter(lambda p: p.name == name, m.persons))
+    if len(found_persons) > 0:
+        return found_persons[0]  # if a person exists, return it
+    else:
+        mm = get_metamodel(m)  # else, create it and store it in the model
+        person = mm['Person']()
```
Comment (goto40) [inline tests/functional/regressions/test_issue167.py]: Is this ok?
Is it ok to instantiate a model object that way? We could maybe add something in the docs for that...

(also check the next lines, where "name" and "parent" are set...

**Comment by goto40** [comment]
(see comment in #167)

**Comment by goto40** [comment]
Just one unittest. I think it is a very valuable addition motivated from @aranega.
If you have no further comments (e.g., on creating the model instance, see my inline comment), this PR could be squashed...

**Comment by igordejanovic** [comment]
@goto40 Yes, this test is really good. But, I wouldn't put it in regression tests as it is not a bug but demonstrates a feature that was already there but haven't been though of until @aranega asked for. 
I suggest to move this test to core scoping tests and add the part of the docs which shortly explains the use-case and links it to the test (and the issue that motivated it).

File: tests/functional/regressions/test_issue167.py (hunk 1 above)
Comment (igordejanovic) [reply tests/functional/regressions/test_issue167.py]: I think this style of instantiation is perfectly fine.

**Review by igordejanovic — CHANGES_REQUESTED** [review]
Please see comments.

**Comment by goto40** [comment]
I do it soon.

**Comment by goto40** [comment]
I moved the test and I wrote some docu. You could have a look. I will be happy to add more aspects and correct things (comments welcome).

File: docs/scoping.md [hunk 2]
```diff
@@ -228,3 +228,43 @@ references to be resolved until all references are resolved or no progress
...
+a reference", you must take into account that these elements
+may have not yet been created. This can be achieved in the same 
+way as handling unresolved references in a scope provider (with the 
+```Postponed``` mechanism). This use case was motivated by 
+[#167](https://github.com/textX/textX/issues/167).
+
+An example of such a meta model is given in 
+[tests/test_model_modification_through_scoping.py](https://github.com/textX/textX/blob/master/tests/functional/test_scoping/test_model_modification_through_scoping.py):
+Here you can 
+ * **define** Persons explicitly (existence) and
```
Comment (igordejanovic) [inline docs/scoping.md:285]: After `Here you can` and additional newline is needed to render the bullets list properly.

File: docs/scoping.md [hunk 3]
```diff
@@ -228,3 +228,43 @@ references to be resolved until all references are resolved or no progress
...
+ * **reference** two persons which **know** each other (relationship). 
+ Moreover, **referencing a not existent persons** (all person explicitly defined
```
Comment (igordejanovic) [inline docs/scoping.md]: `not existent` -> `nonexistent` ?  There are several instances of it.

**Review by igordejanovic — APPROVED** [review]
@goto40 Great work! I really like the test example.
There are just some typos I noticed (please, see inline comments).

Please squash merge when ready.

**Comment by igordejanovic** [comment]
BTW, we could introduce the practice of updating the CHANGELOG with each PR (as a part of it). It would make it easier when the time of release comes, and it would make apparent at any time what is merged to the master branch.

**Comment by goto40** [comment]
**Typos fixed**.

Sorry for updating some more:
I added some text between 

 * "Using the scope provider to modify a model" 
 * and "Use case: reference data "defined by references".

Maybe you can have look again before merging... It is some architectural background explaining the relationship between parsing/reference resolution/object processors...

Maybe you could have a quick look at the additional paragraph...

**Comment by goto40** [comment]
I will push the changelog changes

**Comment by goto40** [comment]
I squash and merge after your "ok". Take your time... I appreciate more comments of course.

**Comment by goto40** [comment]
I will merge now. We can reopen a new ticket later. My additional changes were not so relevant.

**Pull request merged**